### PR TITLE
bpo-44289: Keep argument file object's current position in tarfile.is_tarfile

### PR DESCRIPTION
**Pull request by amateja**
Since Python 3.9 tarfile.is_tarfile accepts not only paths but also files and file-like objects #18090.

Verification if a file or file-like object is a tar file modifies file object's current position.

https://bugs.python.org/issue44289

https://bugs.python.org/issue44289

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@amateja

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by akulakov** [comment]
@amateja hi, you need to sign the CLA..

Other than that, I've looked at the patch and it looks good to me. I've also confirmed that the fix does what it's supposed to by running test with & w/out the fix.

Since this is a (minor) backwards compatibility change, I'm not sure if it needs to also be added to "what's new", I hope @ethanfurman  can comment.

**Comment by amateja** [comment]
Hi @akulakov, thanks for reviewing.

I believe I've already signed CLA. I confirmed that with https://check-python-cla.herokuapp.com/.

Regarding "what's new" I entirely base on your judgment. Just let me know what should be done to unblock this change from being accepted.

**Comment by akulakov** [comment]
@amateja You are right, I somehow missed the CLA accepted message. You're good as far as that goes :-)

**Comment by akulakov** [comment]
I looked at it again and I think this should be fixed in `open()` rather than in `is_tarfile()`, because is_tarfile uses open() so it should be fixed at the source; and because there's already logic in open() that partly addresses this issue:

https://github.com/python/cpython/blob/9fe0de28bcc8d6b4077d9c3994648049730ee721/Lib/tarfile.py#L1616

So the fix would be (a few lines above the linked line; I tested it & it works):

```
                    fp = func(name, "r", fileobj, **kwargs)
                    if fileobj is not None:
                        fileobj.seek(saved_pos)
                    return fp
```

**Comment by akulakov** [comment]
Note also that the fix I suggested covers more issues: it also covers calling `open()` more than a single time, e.g.:

    kwargs = {'fileobj' if hasattr(archive, 'read') else 'name': archive}
    t = tarfile.open(**kwargs)
    t = tarfile.open(**kwargs)
    return [member.name for member in t.getmembers()]

**Comment by amateja** [comment]
Thanks for your review @akulakov. You are right that this should be fixed at the source. I focused myself on `is_tarfile()` because new functionality for file-like objects handling was introduced only there with #18090. But I'm happy to change this approach.

**Comment by akulakov** [comment]
Thanks for making the changes!

I will do a bit of testing tomorrow and look over it again; I think it's enough to have the test that checks that open keeps the same position as you have it, because both bugs are effectively covered by this test.

I will try to post an update tomorrow or in the next 2 days.

**Comment by akulakov** [comment]
I did some profiling to make sure an extra seek doesn't affect performance, I did it by opening 1000 small tar files; here's what I got:

with patch: 415, 417, 418 msec per loop
without patch: 414, 419, 417 msec

As expected, it doesn't make noticeable difference.

```
2850 (~/temp2/temp) % ~/opensource/cpython2/python.exe -m timeit -c 'import tarfile' 'for x in range(1000): tarfile.open("./b%s.tar"%x)'
1 loop, best of 5: 415 msec per loop
2850 (~/temp2/temp) % ~/opensource/cpython2/python.exe -m timeit -c 'import tarfile' 'for x in range(1000): tarfile.open("./b%s.tar"%x)'
1 loop, best of 5: 417 msec per loop
2850 (~/temp2/temp) % ~/opensource/cpython2/python.exe -m timeit -c 'import tarfile' 'for x in range(1000): tarfile.open("./b%s.tar"%x)'
1 loop, best of 5: 418 msec per loop
2850 (~/temp2/temp) % r                                      --INS--
~/opensource/cpython2/python.exe -m timeit -c 'import tarfile' 'for x in range(1000): tarfile.open("./b%s.tar"%x)'
1 loop, best of 5: 414 msec per loop
2852 (~/temp2/temp) % r                                      --INS--
~/opensource/cpython2/python.exe -m timeit -c 'import tarfile' 'for x in range(1000): tarfile.open("./b%s.tar"%x)'
1 loop, best of 5: 419 msec per loop
2854 (~/temp2/temp) % r                                      --INS--
~/opensource/cpython2/python.exe -m timeit -c 'import tarfile' 'for x in range(1000): tarfile.open("./b%s.tar"%x)'
1 loop, best of 5: 417 msec per loop
```

**Comment by akulakov** [comment]
Please update the PR title and the news entry.

For the unit tests, keep both of them for now. A core developer will need to review the PR, so I would wait for them to make this call.

**Comment by akulakov** [comment]
I'm not a bit concerned that users may be relying on the initial seek of `open()`. I will read about the tar spec, perhaps there's a clean way to keep the initial seek but leave it for additional open()'s.

**Comment by akulakov** [comment]
I do think it's plausible that users may be relying on `open()` seeking to 512 bytes, so to minimize the backwards incompatibility change, it's best to allow the first open() to go to 512 bytes, but subsequent open() will reset to 0 first and then go to 512 again.

If users are relying on multiple opens to seek further, they are relying on buggy behaviour  and IMO it's ok to fix that.

Note that when open fails, we still always want to restore original position. It's only on success that  special casing of 512 position comes into play.

I did some testing and the following logic works:
```
                try:
                    if fileobj is not None and saved_pos==BLOCKSIZE:
                        fileobj.seek(0)
                    return func(name, "r", fileobj, **kwargs)
```

It's a bit ugly to hardcode the initial seek position like this, it may be someone more familiar with module internals will suggest a better fix.

**Comment by akulakov** [comment]
Andrzej: I have some spare time right now, if you prefer I can finish the PR and list you as a co-author (because I would have to make it into a new PR). If you are still planning to finish this, no rush.

**Comment by amateja** [comment]
I'm more than happy to work further on this case. I thought we are waiting for someone more familiar with tar internals to comment on your proposal.

**Comment by amateja** [comment]
Hi @akulakov,

I've read some GNU tar manual to understand your proposal better and followed your suggestion with some cosmetic tuning.

Your proposal
```python
                if fileobj is not None:
                    saved_pos = fileobj.tell()
                try:
                    if fileobj is not None and saved_pos == BLOCKSIZE:
                        fileobj.seek(0)
                    return func(name, "r", fileobj, **kwargs)
```

and mine
```python
                if fileobj is not None:
                    saved_pos = fileobj.tell()
                    if saved_pos == BLOCKSIZE:
                        fileobj.seek(0)
                try:
                    return func(name, "r", fileobj, **kwargs)
```

Your fix is just great for uncompressed archives but it does not work for any supported compression methods. Each compression method changes `fileobj` position to value other than 512 bytes (in my tests it was in range 121-529904 depending on compression type and archived files size).

Following case brought me here:
```pycon
>>> import tarfile
>>> 
>>> with open('spam', 'rb') as fileobj:
...     # here fileobj.tell() == 0
...     if tarfile.is_tarfile(fileobj):
...         # here fileobj.tell() != 0
...         archive = tarfile.open(fileobj=fileobj)
...         # therefore listing archive fails
...         print(archive.getmembers())
... 
[]
```
and I have some gut feeling that we should not modify `fileobj` position or even restore it because compressors make a lot of harm to `fileobj` position making above scenario to fail. So I totally understand your reasoning and yet I think we should stick to current fix.

Please use my hacky test method for `CommonReadTest` class if you'd like to verify this on your own.
```python
    def test_is_tarfile_affects_getmembers(self):
        with tarfile.open(name=self.tarname) as f:
            refer = len(f.getmembers())
        with open(self.tarname, "rb") as fileobj:
            self.assertTrue(tarfile.is_tarfile(fileobj))
            pos = fileobj.tell()
            with tarfile.open(fileobj=fileobj) as tar:
                num_files = len(tar.getmembers())
                self.assertEqual(num_files, refer, pos)
```
I'm open for some other proposals and really looking forward to your feedback.

**Comment by akulakov** [comment]
Andrzej: good catch about compressed files, my fix won't work for them.

I think it's not clear if breaking backwards compatibility here is worth it.

One other idea we can explore is setting the saved position on a fileobj and then checking for it on `open` when there's a fileobj. `hasattr` is an expensive operation but here we deal with IO of files which should be much more expensive. One thing I'm not sure if it's possible to set attributes on all types of file objects.

What do you think?

**Comment by akulakov** [comment]
Andrzej: can you also check if on subsequent `open`s, the seek is always a multiple of first seek? That was the case with uncompressed.

If so, we can do logic like: `x=current pos; open(); y=current pos; if y==x*2: seek to x`

**Comment by amateja** [comment]
Hi Andrei,

Let me start with answering your question. It turns out that calling subsequently `open` moves pointer forward by random number of bytes:

```pycon
>>> import tarfile
>>> name = 'spam.tar.gz'
>>> f = open(name, 'rb')
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b8d4fd0>
>>> f.tell()
8202
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b9eb0a0>
>>> f.tell()
8714
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b8d4fd0>
>>> f.tell()
9226
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7bb86d60>
>>> f.tell()
9738
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b9eb220>
>>> f.tell()
10250
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b9d4fa0>
>>> f.tell()
10762
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b8d4fd0>
>>> f.tell()
11274
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7bb86d60>
>>> f.tell()
11786
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7b9eb0a0>
>>> f.tell()
12298
>>> tarfile.open(fileobj=f)
<tarfile.TarFile object at 0x7fae7bb86ac0>
>>> f.tell()
12810
```
So for now we have few options here:
1. look for unclear and hacky way to fix this issue keeping most of backward compatibility
   This is not very easy as compressed files behaves oddly and we are making this module over-complicated.
2. fall back to my original proposal with fix in `is_tarfile` function only
   Not the best but at least it covers quite reasonable scenario - check if a file-like object is a tar file and open if it is.
3. disregard this issue completely
   I can live without it and most of our community will. We are dealing with checking / opening tar files differently depending if they are described by name as a string of file-like object. This is an inconsistency to me but we can cover this with some mention in the documentation.
4. fix it completely with relevant documentation
   Slightly rough but I don't see those use cases you wrote about. I cannot imagine why someone might rely on modified state of file-like object argument. I'm sure there might be some good reason for that. You wrote that multiple `open`s to seek further is buggy behavior and I agree. Knowing that compressed files behaves even more unexpectedly I would say no one should use it that way. I verified that comparing pointer's position in multiple `open`s vs. calling `Tarfile.next()` and is unrelated so this is not it.

I like option 4 and option 2 is not too shabby either. I would also back option 3 up if there is no other way and I'm not very fond of option 1 in it's current state.

**Comment by amateja** [comment]
Oh wait... I did some more math. First `open` moves pointer to some location but second and next `open`s moves it forward by 512 bytes again. So at least we know this "unclear and hacky way".

**Comment by akulakov** [comment]
In my testing the pointer doesn't always advance, with two tiny text files, gzip format, it advanced from 0 to 202 but with subsequent `open()`'s, it stayed at 202.

One additional option we have is to wait for more user feedback and for somebody to come up with a better fix. I will think more about this later tonight / tomorrow.

**Comment by akulakov** [comment]
Andrzej: after some discussion on discord, I think the option 2) (your original solution), is the best way to address this.

**Comment by amateja** [comment]
Aye aye. Please find recent commit reverted.

File: Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst
Comment (akulakov) [inline Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst]: News entry can be more descriptive, so that users can easily tell if they need to update their codebase;
something along these lines:

Fix an issue with :meth:`~tarfile.is_tarfile` method when using *fileobj* argument: position
in the *fileobj* was advanced forward which made it unreadable with :meth:`tarfile.TarFile.open`.

Comment (amateja) [reply Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst]: I tried to come up with something for new entry but your version is just right. I hope you don't mind me using it, do you?

Comment (akulakov) [reply Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst]: Of course, go ahead!

File: Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst
Comment (akulakov) [inline Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst]: In documentation, bold highlight (with asterisks) is generally used for parameter names, for example see https://github.com/python/cpython/edit/main/Doc/whatsnew/3.10.rst line 913

**Comment by akulakov** [comment]
Aside from one small nitpick in the news file, LGTM! 
I also tested with a tarfile locally and it works.

@gvanrossum As you looked at it a couple of days ago, maybe you can review?

**Review by gvanrossum — APPROVED** [review]
Yeah, this is what I expected it would look like. Thanks for the thorough tests! I will merge this now.

**Comment by gvanrossum** [comment]
(Whoops, let's run the tests before merging rashly. :-)

File: Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst
Comment (amateja) [reply Misc/NEWS.d/next/Library/2021-06-02-19-47-46.bpo-44289.xC5kuV.rst]: Good catch. Thanks. I've also noticed that here and there monospace is used and I'm not sure why.

https://github.com/python/cpython/blame/main/Doc/whatsnew/3.10.rst#L994
https://github.com/python/cpython/blame/main/Doc/whatsnew/3.10.rst#L1015

Fixed already. Happy to follow guidelines.

**Review by kumaraditya303 — APPROVED** [review]
LGTM, the PR title has a typo which should be fixed before merging.

**Comment by gvanrossum** [comment]
(And the tests need to run one more time.)

**Comment by amateja** [comment]
Many thanks to @akulakov for going together through this issue with me. Thank you @gvanrossum for reviewing and @kumaraditya303 for approving :)

Looking forward to working with you again.

**Comment by akulakov** [comment]
Andrzej: you're very welcome, thanks for contributing!

**Pull request merged**
